lib/rust/flexer/src/group.rs

```diff
@@ -111,120 +225,131 @@ impl From<&Group> for NFA {
 pub mod tests {
     extern crate test;
 
-    use crate::automata::dfa::DFA;
     use crate::automata::nfa;
-    use crate::automata::nfa::NFA;
     use crate::automata::pattern::Pattern;
-    use crate::group::Group;
+    use crate::group::{Group, GroupRegistry};
     use crate::group::rule::Rule;
 
     use std::default::Default;
     use test::Bencher;
+    use enso_prelude::default;
 
-    fn newline() -> Group {
+    fn newline() -> GroupRegistry {
         let     pattern = Pattern::char('\n');
         let mut group   = Group::default();
 
         group.add_rule(Rule{pattern,callback:"".into()});
 
-        group
+        let mut registry = GroupRegistry::default();
+        registry.add_group(group);
+        registry
     }
 
-    fn letter() -> Group {
+    fn letter() -> GroupRegistry {
         let     pattern = Pattern::range('a'..='z');
```

Comment:
instead of these `Pattern::range` and `Pattern::char` etc you should also be able to abstract it away and write just `Pattern::from` and create from impls for every possible constructor input.

Reply:
I think this would be less clear as there are multiple types of pattern that can take a given type of input.